### PR TITLE
SCP-2638: simplify datatypes which are used only at the type-level

### LINKED ISSUE #3702: "Empty" validator generates 3KB serialized contracts.
## Area

[x] *Plutus Foundation* Related to the GHC plugin, Haskell-to-Plutus compiler, on-chain code
[] *Plutus Application Framework* Related to the Plutus application backend (PAB), emulator, Plutus libraries
[] *Marlowe* Related to Marlowe
[] *Other* Any other topic (Playgrounds, etc.)

## Summary

Empty validators generates serialized scripts of 3KB filled with unused primitives. Perhaps expected at this stage? 

## Steps to reproduce

Out of curiosity and to better understand where some errors in another contract could com from, I checked the size of an empty validator:

```hs
data Initial

instance Scripts.ValidatorTypes Initial where
  type DatumType Initial = ()
  type RedeemerType Initial = ()

validator :: () -> () -> ScriptContext -> Bool
validator _ _ _ = True

compiledValidator :: CompiledCode (ValidatorType Initial)
compiledValidator = $$(compile [||validator||])
```

The resulting compiled version is ~3KB and looking at the pretty-printed compiled code (see below), it seems mostly filled with primitives and base constructions for the `ScriptContext`. Even if unused. 

<details>
    <summary>see compiled code</summary>

```
(program
  (let
    (nonrec)
    (datatypebind
      (datatype (tyvardecl Unit (type))  Unit_match (vardecl Unit Unit))
    )
    (datatypebind
      (datatype
        (tyvardecl Credential (type))

        Credential_match
        (vardecl PubKeyCredential (fun (con bytestring) Credential))
        (vardecl ScriptCredential (fun (con bytestring) Credential))
      )
    )
    (datatypebind
      (datatype
        (tyvardecl StakingCredential (type))

        StakingCredential_match
        (vardecl StakingHash (fun Credential StakingCredential))
        (vardecl
          StakingPtr
          (fun (con integer) (fun (con integer) (fun (con integer) StakingCredential)))
        )
      )
    )
    (datatypebind
      (datatype
        (tyvardecl DCert (type))

        DCert_match
        (vardecl DCertDelegDeRegKey (fun StakingCredential DCert))
        (vardecl
          DCertDelegDelegate
          (fun StakingCredential (fun (con bytestring) DCert))
        )
        (vardecl DCertDelegRegKey (fun StakingCredential DCert))
        (vardecl DCertGenesis DCert)
        (vardecl DCertMir DCert)
        (vardecl
          DCertPoolRegister (fun (con bytestring) (fun (con bytestring) DCert))
        )
        (vardecl
          DCertPoolRetire (fun (con bytestring) (fun (con integer) DCert))
        )
      )
    )
    (datatypebind
      (datatype
        (tyvardecl TxOutRef (type))

        TxOutRef_match
        (vardecl TxOutRef (fun (con bytestring) (fun (con integer) TxOutRef)))
      )
    )
    (datatypebind
      (datatype
        (tyvardecl ScriptPurpose (type))

        ScriptPurpose_match
        (vardecl Certifying (fun DCert ScriptPurpose))
        (vardecl Minting (fun (con bytestring) ScriptPurpose))
        (vardecl Rewarding (fun StakingCredential ScriptPurpose))
        (vardecl Spending (fun TxOutRef ScriptPurpose))
      )
    )
    (datatypebind
      (datatype
        (tyvardecl Bool (type))

        Bool_match
        (vardecl True Bool) (vardecl False Bool)
      )
    )
    (datatypebind
      (datatype
        (tyvardecl Extended (fun (type) (type)))
        (tyvardecl a (type))
        Extended_match
        (vardecl Finite (fun a [Extended a]))
        (vardecl NegInf [Extended a])
        (vardecl PosInf [Extended a])
      )
    )
    (datatypebind
      (datatype
        (tyvardecl LowerBound (fun (type) (type)))
        (tyvardecl a (type))
        LowerBound_match
        (vardecl LowerBound (fun [Extended a] (fun Bool [LowerBound a])))
      )
    )
    (datatypebind
      (datatype
        (tyvardecl UpperBound (fun (type) (type)))
        (tyvardecl a (type))
        UpperBound_match
        (vardecl UpperBound (fun [Extended a] (fun Bool [UpperBound a])))
      )
    )
    (datatypebind
      (datatype
        (tyvardecl Interval (fun (type) (type)))
        (tyvardecl a (type))
        Interval_match
        (vardecl Interval (fun [LowerBound a] (fun [UpperBound a] [Interval a]))
        )
      )
    )
    (datatypebind
      (datatype
        (tyvardecl Maybe (fun (type) (type)))
        (tyvardecl a (type))
        Maybe_match
        (vardecl Just (fun a [Maybe a])) (vardecl Nothing [Maybe a])
      )
    )
    (datatypebind
      (datatype
        (tyvardecl Address (type))

        Address_match
        (vardecl
          Address (fun Credential (fun [Maybe StakingCredential] Address))
        )
      )
    )
    (datatypebind
      (datatype
        (tyvardecl Tuple2 (fun (type) (fun (type) (type))))
        (tyvardecl a (type)) (tyvardecl b (type))
        Tuple2_match
        (vardecl Tuple2 (fun a (fun b [[Tuple2 a] b])))
      )
    )
    (let
      (rec)
      (datatypebind
        (datatype
          (tyvardecl List (fun (type) (type)))
          (tyvardecl a (type))
          Nil_match
          (vardecl Nil [List a]) (vardecl Cons (fun a (fun [List a] [List a])))
        )
      )
      (let
        (nonrec)
        (datatypebind
          (datatype
            (tyvardecl TxOut (type))

            TxOut_match
            (vardecl
              TxOut
              (fun Address (fun [[(lam k (type) (lam v (type) [List [[Tuple2 k] v]])) (con bytestring)] [[(lam k (type) (lam v (type) [List [[Tuple2 k] v]])) (con bytestring)] (con integer)]] (fun [Maybe (con bytestring)] TxOut)))
            )
          )
        )
        (datatypebind
          (datatype
            (tyvardecl TxInInfo (type))

            TxInInfo_match
            (vardecl TxInInfo (fun TxOutRef (fun TxOut TxInInfo)))
          )
        )
        (datatypebind
          (datatype
            (tyvardecl TxInfo (type))

            TxInfo_match
            (vardecl
              TxInfo
              (fun [List TxInInfo] (fun [List TxOut] (fun [[(lam k (type) (lam v (type) [List [[Tuple2 k] v]])) (con bytestring)] [[(lam k (type) (lam v (type) [List [[Tuple2 k] v]])) (con bytestring)] (con integer)]] (fun [[(lam k (type) (lam v (type) [List [[Tuple2 k] v]])) (con bytestring)] [[(lam k (type) (lam v (type) [List [[Tuple2 k] v]])) (con bytestring)] (con integer)]] (fun [List DCert] (fun [List [[Tuple2 StakingCredential] (con integer)]] (fun [Interval (con integer)] (fun [List (con bytestring)] (fun [List [[Tuple2 (con bytestring)] (con data)]] (fun (con bytestring) TxInfo))))))))))
            )
          )
        )
        (datatypebind
          (datatype
            (tyvardecl ScriptContext (type))

            ScriptContext_match
            (vardecl
              ScriptContext (fun TxInfo (fun ScriptPurpose ScriptContext))
            )
          )
        )
        (termbind
          (strict)
          (vardecl validator (fun Unit (fun Unit (fun ScriptContext Bool))))
          (lam ds Unit (lam ds Unit (lam ds ScriptContext True)))
        )
        validator
      )
    )
  )
)
```

</details>

As a side note, what I find also interesting is that the pretty-printed code is ~5KB, which isn't that much bigger than the serialized code. Intuitively, I'd expect the serialized code to be at least an order of magnitude smaller :thinking:. Plus, doing 'dummy' transformations on the pretty-printed code like replacing a few of the keywords by some shorter identifier (e.g. vardecl -> v, datatypebind -> dtb, fun -> f ...) rapidly brings down the size of the pretty-printed code to 3KB and less, which suggests that something is wrong with the serialized code?

## Expected behavior

- The compiled code strips out dead-code and does not include unnecessary code.
- The serialized code is more compact.

## System info (please complete the following information):

- OS: Linux / Manjaro
- Version: N/A
- Plutus version or commit hash: 
  ```
  commit 35b1e41b162a5bae87955a86e5eea33535130ff9
  Author: Sjoerd Visscher <sjoerd.visscher@iohk.io>
  Date:   Wed Jul 28 13:41:51 2021 +0200
  ```

## Screenshots and attachments

- If applicable, add screenshots, config files and/or logs to help explain the problem.
- If the issue is related to the Plutus Playground, please include a link to a gist that reproduces the issue

### LINKED ISSUE #4147: Avoid retaining datatypes in PIR which are used only at the type level
We currently retain datatypes which are used only at the type level, including their constructors, matchers, and datatypes which they depend on.

This is true even if the datatype is mentioned only as the type of an ignored argument, for example. It would be nice not to do that: as a crude measure, we could replace them with any made up type with the same kind.

------------------------

Example (in pseudo-PIR):
```
let datatype (Maybe :: Type -> Type) [a :: Type] match_Maybe [Just :: a -> Maybe a, Nothing :: Maybe a]
in \(x :: Maybe Integer) -> \(y :: Integer) -> y
```
Here `Maybe` appears only at the type level. We would like to get rid of as much of `Maybe` as we can (in particular, that will then lead to cascading opportunities to get rid of other stuff), but we do want to keep the program typechecking.

My idea is to replace `Maybe` with a simple made-up type of kind `Type -> Type`. In this case we can use a trivial type-lambda:
```
let Maybe :: Type -> Type = \(t :: Type) -> t
in \(x :: Maybe Integer) -> \(y :: Integer) -> y
```

More generally, datatypes are going to tend to have types that look like a tree of `(->)`s with leaves of kind `Type`. It should be pretty easy to generate a "simple" type of the same kind just using type lambdas.

Then we just replace the datatype binding for `Maybe` with a simple type binding with the new made-up type, and dead-code elimination should do the rest.

-----------------

Working out which types are only retained in this way may be a bit annoying: the dead code analysis has no notion of "provenance". And we don't want to treat these as "dead", rather we need to identify them separately. Possibly the thing to do would be to run the dead code analysis again in a separate mode that ignores dependencies from terms to types, and then just look at what datatypes look "dead" in that analysis. Or maybe we could do something cleverer.

### LINKED ISSUE #4414: add support for lists, pairs and data to concrete syntax
### Describe the feature you'd like

The parser and pretty printer do not properly support lists, pairs and data.

I would like to use the syntax

```con (list bool) [True, False, True]```
and ```con (list (pair integer bool)) [(1, True), (500000, False), (0, True)]```
for lists and pairs.

@SchmErik proposes the following syntax for `data` which I think is nice:
```
(program 0.0.0 (con data {
    Map [ (ByteString #0123,   Integer 12345),
          (ByteString #456789, Integer 789453),
          (ByteString #0ABCDE, Integer 12364689486)]
}))
```

Note that the concrete syntax is not used on chain, is not guaranteed to be production quality, and is subject to change. For testing production code, the flat representation should be used. Concrete syntax is useful for pedagogical and exploratory purposes.

### Describe alternatives you've considered

I suppose one could write something more json like inside the curly brackets for data.

### PR DESCRIPTION
**Pull request by michaelpj**
I got annoyed with what I was doing and I wanted to polish off something where I knew what to do, so I did this. cc @kk-hainq also.

------------

1. Change dependency analysis to account for the fact that the
term-level parts can be removed (see note).
2. Simplify datatype bindings into trivial type bindings if all their
term-level parts are dead.

Had to do a bit of test rearrangement since a lot of the
`plutus-tx-plugin` tests for a type T just used a lambda with an unused
argument of type T... which gets simplified with this PR!

Fixes #4147.
Fixes #3702.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Comment by kk-hainq** [comment]
You got the wrong issue number again! 

Anyhow, looks like something I should have done given your past suggestions, cannot wait to get back on track with these works soon :+1:.

**Comment by michaelpj** [comment]
Hydra is being weird, this passed fine.

File: plutus-tx-plugin/test/Plugin/Data/Spec.hs [hunk 1]
```diff
@@ -69,8 +69,8 @@ instance P.Eq MyMonoData where
     (Mono3 i1) == (Mono3 i2)       = i1 P.== i2
```
Comment (bezirg) [inline plutus-tx-plugin/test/Plugin/Data/Spec.hs:69]: Perhaps add a coment in all of these that says that we have to patternmatch, so we don't optimize away the `data => type`?

**Review by bezirg — APPROVED** [review]

File: plutus-core/plutus-ir/src/PlutusIR/Transform/DeadCode.hs [hunk 2]
```diff
@@ -55,24 +57,51 @@ live n =
...
+            constrsLive <- traverse liveVarDecl constrs
+            let termLive = or (destrLive : constrsLive)
+            case (dtypeLive, termLive) of
+                 -- At least one term-level part is live, keep the whole thing
+                (_, True)      -> pure $ Just b
+                -- Nothing is live, remove the whole thing
+                (False, False) -> pure Nothing
+                -- See Note [Dependencies for datatype bindings, and pruning them]
+                 -- Datatype is live but no term-level parts are, replace with a trivial type binding
+                (True, False)  -> Just . TypeBind x d <$> mkTypeOfKind (_tyVarDeclKind d)
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/DeadCode.hs:87]: Yeah, I actually thought about this! But I didn't think it was clearer in the end. In particular, I was only doing one `guard` in each control flow path, and only at the end. I think if you're *also* chaining the `Maybe`s it's useful, but here it was just clutter.

File: plutus-tx-plugin/test/Plugin/Data/Spec.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-tx-plugin/test/Plugin/Data/Spec.hs:69]: :+1:

File: plutus-core/plutus-ir/src/PlutusIR/Analysis/Dependencies.hs [hunk 3]
```diff
@@ -137,16 +164,19 @@ bindingDeps b = case b of
...
         tvDeps <- traverse tyVarDeclDeps tvs
         cstrDeps <- traverse varDeclDeps constrs
-        -- All the datatype bindings depend on each other since they can't be used separately. Consider
-        -- the identity function on a datatype type - it only uses the type variable, but the whole definition
-        -- will therefore be kept, and so we must consider any uses in e.g. the constructors as live.
-        let tyus = fmap (view PLC.theUnique) $ _tyVarDeclName d : fmap _tyVarDeclName tvs
-        let tus = fmap (view PLC.theUnique) $ destr : fmap _varDeclName constrs
-        let localDeps = G.clique (fmap Variable $ tyus ++ tus)
-        pure $ G.overlays $ [vDeps] ++ tvDeps ++ cstrDeps ++ [localDeps]
+        -- Destructors depend on the datatype and the argument types of all the constructors.
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Analysis/Dependencies.hs:171]: Nope, I meant what I said!

[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Analysis/Dependencies.hs:171]: I'll elaborate.

**Review by thealmarty — APPROVED** [review]
I don't think I got a deep understanding of this......but I noticed that some of the golden files are longer, not shorter. I thought the goal is to remove more dead code?

**Comment by michaelpj** [comment]
> I noticed that some of the golden files are longer, not shorter. I thought the goal is to remove more dead code?

That's because I changed the inputs! As I alluded to in the PR description, I had to change a few tests to avoid the new optimization removing stuff that the test actually cared about!

**Review by ak3n — APPROVED** [review]

**Pull request merged**